### ISSUE TITLE
introduce nightly build?

### ISSUE THREAD

**Issue by robstoll** [robstoll/atrium-roadmap#31]
shall we introduce a dev version published every night?

in case we do, then I suggest that we also publish a nightly kdoc

**Comment by jGleitz** [comment]
If we do, I think it should not be a nightly build, but rather an automation that publishes the latest master on push. 

Currently, atrium does not release very often. Because of that, having a development build would enable users to start using the latest features without having to wait for you to make a release.

On the other hand, it might make sense to try and release a lot more often (roughly after every PR that adds or fixes functionality) so users can profit from improvements as fast as possible and without having to resort to development builds.

One does not contradict the other, of course.

**Comment by robstoll** [comment]
> If we do, I think it should not be a nightly build, but rather an automation that publishes the latest master on push.

The problem with publishing after each push on master is that pushes can happen very frequently and publishing takes unfortunately ~30min to bintray and roughly the same time to sync to maven central.
IMO once per night would be enough as most probably users don't want to change the dependency more often than once per day. 
However, if someone waits for a particular PR being published, waiting till the next day might be too long as well. 
We could do something like, publish automatically if the last publish is older than 2h and check once per night if there have been new commits which where inside this 2h window and thus could be published as well.

> On the other hand, it might make sense to try and release a lot more often (roughly after every PR that adds or fixes functionality) 

basically I agree.... but not until I have the feeling we are more stable and more or less add only new features. We could of course publish -alpha1, -alpha2 etc. but then we can also use SNAPSHOTS IMO

**Comment by jGleitz** [comment]
If publishing takes that long, we could also publish to GitHub packages, which is pretty fast (but requires a token to pull from).

Or we could just confirm that this repo works with [JitPack](https://jitpack.io/). Users could then use whatever git ref they are interested in.

**Comment by robstoll** [comment]
publishing to Bintray requires a token as well (if done automatically). I wasn't aware of JitPack, sounds interesting as well

btw. I think ~15min is used by dokka

**Comment by jGleitz** [comment]
> publishing to Bintray requires a token as well

yes, but _pulling from_ GitHub packages also requires a token!

**Comment by robstoll** [comment]
Ah really, cool. Might use it for other purposes then 😄

**Comment by jGleitz** [comment]
Renovate uses the  [semantic-release](https://github.com/semantic-release/semantic-release) bot and I really like the concept. The basic idea is that commits (or PR titles, but then the PR needs to be squashed) contain information about what the commit/PR does. The bot then calculates the appropriate version number and releases automatically. It even posts on the issue which version number fixes it.

The consequence would be that we need to teach contributors to use semantic commits. But since we can also just change the PR title and squash, it should not be a big issue.

Regarding

> The problem with publishing after each push on master is that pushes can happen very frequently and publishing takes unfortunately ~30min to bintray and roughly the same time to sync to maven central.

 a) I don’t think that the publishing delay is an argument against publishing more often. It’s not like we pay for that, do we?
b) There are not that many PRs. I think the frequency of master commits mainly comes from your activity. If you started doing PRs as well, this would also give more transparency regarding what changes in Atrium. it would also give contributors a chance to voice their opinions on changes (I would not make reviews mandatory, though). What are your thoughts on that?

**Comment by robstoll** [comment]
a) no, we don't pay. I was more thinking that people using bots to update dependencies won't be very happy to get a new PR every hour in case I am working on things and push regularly to master :)
Also, I think things like `fix typo in doc` do not really deserve a new release. But maybe you were talking about publishing dev versions more often and I am mixing things up here. I have nothing against publishing dev versions more often.

b) I usually work on feature branches and do merge commits. I could go over PRs but this extra step would only make sense if someone reviews. Yet, I would still merge the PR myself as I don't want to wait and be blocked, so it would be more like an after-review and suggestions could be applied in a future PR. It could be a good way to get more into Atrium  for regular contributors. Would you like to review things?

I don't want to use the semantic-release bot right now but maybe something for later on. Feel free to create another issue.

**Comment by jGleitz** [comment]
> Also, I think things like fix typo in doc do not really deserve a new release.

I fully agree. I initially thought about publishing only dev versions frequently, however, I have changed my mind about it. In fact, I now propose to push a new release whenever there is a new feature. semantic-release would be one solution to achieve this because it gives fine-grade control over when to publish.

> Yet, I would still merge the PR myself as I don't want to wait and be blocked

Definitely! The process should not block you.

> Would you like to review things?

I would. I probably don’t have the time to review everything. I also don’t know whether I can contribute anything of value. But it would, as you say, give me a chance to better understand what is currently going on in atrium.

I created https://github.com/robstoll/atrium-roadmap/issues/60.
